### ISSUE TITLE
Lifecycle participant - is it working correctly?

### ISSUE THREAD

**Issue by bboyle1234** [dotnet/orleans#4157]
This is a lifecycle participant intended to run `OnStart` before any grains are activate, and `OnStop` after all grains are deactivated.

It seems to correctly run `OnStart` before any grains are activated, but it seems to **incorrectly** run `OnStop` before the grains are deactivated.

```
    public interface ILocalSiloClusterClientProvider {
        IClusterClient Client { get; }
    }

    public class LocalSiloClusterClientProvider : ILifecycleParticipant<ISiloLifecycle>, ILocalSiloClusterClientProvider {

        readonly Action<ClientBuilder> ConfigureClientBuilder;

        public IClusterClient Client { get; private set; }

        public LocalSiloClusterClientProvider(Action<ClientBuilder> configureClientBuilder) {
            ConfigureClientBuilder = configureClientBuilder;
        }

        public void Participate(ISiloLifecycle lifecycle) {
            lifecycle.Subscribe(ServiceLifecycleStage.ApplicationServices, OnStart, OnStop);
        }

        async Task OnStart(CancellationToken token) {
            var clientBuilder = new ClientBuilder();
            ConfigureClientBuilder?.Invoke(clientBuilder);
            Client = clientBuilder.Build();
            await Client.Connect();
        }

        async Task OnStop(CancellationToken token) {
            await Client.Close();
            Client.Dispose();
        }
    }

    public static class LocalSiloClusterClientProviderExtensions {
        public static ISiloHostBuilder AddLocalSiloClusterClientProvider(this ISiloHostBuilder builder, Action<ClientBuilder> clientBuilder) {
            return builder.ConfigureServices(services => {
                services.AddSingleton<ILocalSiloClusterClientProvider, LocalSiloClusterClientProvider>(serviceProvider => new LocalSiloClusterClientProvider(clientBuilder));
                services.Add(new ServiceDescriptor(typeof(ILifecycleParticipant<ISiloLifecycle>), sp => sp.GetRequiredService<ILocalSiloClusterClientProvider>(), services.Last().Lifetime));
            });
        }
    }
```
The service is injected on the SiloBuilder with the following code: 

```
    .AddLocalSiloClusterClientProvider(clientBuilder => clientBuilder
            .ConfigureCluster(options => options.ClusterId = MyGrainConfiguration.ClusterId)
            .ConfigureApplicationParts(parts => parts.AddApplicationPart(typeof(IInstrumentGrain).Assembly).WithReferences())
            .UseStaticClustering(options => { options.Gateways.Add(new IPEndPoint(IPAddress.Loopback, MyGrainConfiguration.GatewayPort).ToGatewayUri()); }))
```

Silo shutdown is initiated via `await Silo.StopAsync()`, while running in a dot net core 2.0 console app after the user presses Ctrl+C.

Silo is also configured with `.Configure<ProcessExitHandlingOptions>(options => options.FastKillOnProcessExit = false)`

```
            Console.CancelKeyPress += (s, a) => {
                a.Cancel = true;
                Task.Run(StopSilo);
            };

        static async Task StopSilo() {
            await silo.StopAsync();
            _siloStopped.Set(); // a manual reset event that allows 'Main' to finish up
        }
```

**Comment by benjaminpetit** [comment]
In your example in your connecting to the silo before it is active. With the recent fixes in Silo startup, it will no longer work as expected (silo will not accept any request before the active state). 

It sees that we deactivate grain in the `RuntimeServices` state, that might be too late...

**Comment by jason-bragg** [comment]
Clients connect to clusters, not silos, so it's not clear when a client on a silo would be able to connect.  For instance, if a silo is joining a running cluster, a client on that silo could connect before the silo is even part of the cluster (very early), but if the silo is the first or only silo in the cluster, the client could not connect until the silo has become active.

If the client is needed prior to the silo joining the cluster, I'd advocate using lazy initialization, and requiring callers to handle the cluster connection errors, rather than participating in the silo lifecycle.

It should also be noted that the building of the client, connecting to the cluster, and disconnecting need not be all done in the same stages.  For example, the building of the client, and disconnection from the cluster, can be performed at an earlier stage of the lifecycle than the connect.

**Comment by bboyle1234** [comment]
Understood that the client will have to connect to the cluster not the localhost.

Lazy client initialization - sounds like a good solution. Orleans shouldn't be accepting client connections until the grains are able to be started. 

I'm 100% behind Orleans not accepting client connections until after Activation state (20000).
Perhaps the client connection could initialize itself at state value 20001? Would that be a practical solution instead of the lazy intialization idea previously floated?

----------------------------------

So grains are not deactivated until RuntimeServices .... I understand why it has been done this way .... Orleans special internal service grains need to stay alive longer. However, I think the average grain should be deactivated before the storage services shutdown. Perhaps ordinary grains should deactivate at stage 20000, and Orleans special grains be marked to deactivate later?

I've written a test project where grains are not calling `await WriteStateAsync()` until they are deactivated. Can you confirm my suspicion that they are trying to write state after the storage services have been shutdown and are no longer working?

**Comment by benjaminpetit** [comment]
> However, I think the average grain should be deactivated before the storage services shutdown. Perhaps ordinary grains should deactivate at stage 20000, and Orleans special grains be marked to deactivate later?
> I've written a test project where grains are not calling await WriteStateAsync() until they are deactivated. Can you confirm my suspicion that they might be trying to write state after the storage services have been shutdown and are no longer working?

You are correct, https://github.com/dotnet/orleans/pull/4177 should fix this :)

**Comment by bboyle1234** [comment]
@benjaminpetit  :)

**Comment by sergeybykov** [comment]
Resolved via #4177.
